hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/lock/WrappedStripedLock.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.om.lock;
+
+import com.google.common.util.concurrent.Striped;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import org.apache.hadoop.hdds.utils.SimpleStriped;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Locking class that wraps Striped locking with timeout and logging. It also provides ordering while multiple locks.
+ */
+public class WrappedStripedLock {
+  private static final Logger LOG = LoggerFactory.getLogger(WrappedStripedLock.class);
+  private final Striped<ReadWriteLock> fileStripedLock;
+  private final long lockTimeout;
+
+  public WrappedStripedLock(int stripLockSize, long timeout, boolean fair) {
+    fileStripedLock = SimpleStriped.readWriteLock(stripLockSize, fair);
+    lockTimeout = timeout;
+  }
+
+  /**
+   * lock the list of keys in order.
+   * Sample code for lock and unlock handling:
+   * <code>
+   *   try {
+   *     if (!wrappedStripedLock.lock(keyList, locks)) {
+   *       // timeout occurred, release lock if any in reverse order
+   *       Collections.reverse(locks);
+   *       locks.forEach(Lock::unlock);
+   *     }
+   *     // perform business logic
+   *   } finally {
+   *     // to be released in reverse order
+   *     Collections.reverse(locks);
+   *     locks.forEach(Lock::unlock);
+   *   }
+   * </code>
+   * 
+   * @param keyList key list which needs to be locked
+   * @param locks successful lock object returned which will be used to release lock
+   * @return boolean true if success, else false
+   * @throws InterruptedException exception on interrupt
+   */
+  public boolean writeLock(List<String> keyList, List<Lock> locks) throws InterruptedException {
+    try {
+      Iterable<ReadWriteLock> readWriteLocks = fileStripedLock.bulkGet(keyList);
```

Comment:
Shouldn't we always sort the keyList before doing a bulkGet here? This could lead to deadlocks otherwise

Reply:
We can have a Map<String, Lock> instead of a list

Reply:
The lock order is determined by index calculated by Stripped lock in bulkGet(). This ordering is based on index from file name. Sorting is redundant and not used as re-ordering will be done by bulkGet().

We need use this ordering to avoid deadlock based on index, and need use List for maintain lock and unlock ordering.